### PR TITLE
Test clearing clipboard behavior instead of implementation

### PR DESCRIPTION
**Pull request by gonzalo-bulnes**
# Description

The application must clear the clipboard when requested. Previously, the test would mock the clipboard and assert on its API being used to clear it (the `QClipboard.clear()` function was called). This PR removes the need for mocking by treating the clipboard like an external dependency, and asserts that its contents are actually cleared when expected, without assumptions about the means the application used to clear them.

Because the test itself doesn't depend on the code implementation, it can be used as a reference to ensure the behavior remains correct as we modify the application.

Example for https://github.com/freedomofpress/securedrop-client/issues/1345

# Test Plan

- [x] The test suite should be green
- [x] Confirm that the test fails if the [corresponding code](https://github.com/freedomofpress/securedrop-client/pull/1346/files#diff-1ac9a12fab465eeb04cbe96918b42f5ab3d9d2fcc003ff6d2dd4d9f78008329aR225) is commented

# Checklist

If these changes modify code paths involving cryptography, the opening of files in VMs or network (via the RPC service) traffic, Qubes testing in the staging environment is required. For fine tuning of the graphical user interface, testing in any environment in Qubes is required. Please check as applicable:

 - [ ] I have tested these changes in the appropriate Qubes environment
 - [ ] I do not have an appropriate Qubes OS workstation set up (the reviewer will need to test these changes)
 - [x] These changes should not need testing in Qubes

If these changes add or remove files other than client code, the AppArmor profile may need to be updated. Please check as applicable:

 - [ ] I have updated the [AppArmor profile](https://github.com/freedomofpress/securedrop-client/blob/HEAD/files/usr.bin.securedrop-client)
 - [x] No update to the AppArmor profile is required for these changes
 - [ ] I don't know and would appreciate guidance

If these changes modify the database schema, you should include a database migration. Please check as applicable:

 - [ ] I have written a migration and upgraded a test database based on `main` and confirmed that the migration applies cleanly
 - [ ] I have written a migration but have not upgraded a test database based on `main` and would like the reviewer to do so
 - [ ] I need help writing a database migration
 - [x] No database schema changes are needed

### CODE REVIEW

File: tests/gui/test_main.py [hunk 1]
```diff
@@ -10,6 +12,18 @@
...
+class WindowTest(unittest.TestCase):
+    def setUp(self):
+        self._clipboard = QApplication.clipboard()
+        self.window = Window(self._clipboard)
+
+    def test_clear_clipboard(self):
+        self._clipboard.setText("The clipboard is not empty")
+
+        self.window.clear_clipboard()
+        assert self.window.clipboard.text() == ""
```
Comment (gonzalo-bulnes) [inline tests/gui/test_main.py:24]:
```suggestion
        assert self._clipboard.text() == ""
```

Oops, this should be the assertion. The fact that the window hold a reference to to clipboard is an implementation detail. I'll fix it tomorrow.

**Review by rocodes — DISMISSED** [review]
LGTM, login credentials in the clipboard are still cleared on login, and tests pass locally.

**Review by creviera — COMMENTED** [review]
As a suggestion, the diff below is a smaller change that improves the test without needing to keep a reference to the application's clipboard in the window.

```diff
diff --git a/tests/gui/test_main.py b/tests/gui/test_main.py
index 318625b..e10a2a4 100644
--- a/tests/gui/test_main.py
+++ b/tests/gui/test_main.py
@@ -363,8 +363,9 @@ def test_clear_clipboard(mocker):
     """
     Ensure we are clearing the system-level clipboard in the expected manner.
     """
-    mock_clipboard = mocker.MagicMock()
-    mocker.patch("securedrop_client.gui.main.QApplication.clipboard", return_value=mock_clipboard)
     w = Window()
+    cb = app.clipboard()
+    cb.setText("the clipboard is not empty")
+    assert cb.text() == "the clipboard is not empty"
     w.clear_clipboard()
-    mock_clipboard.clear.assert_called_once_with()
+    assert cb.text() == ""
```

**Comment by gonzalo-bulnes** [comment]
@creviera Can you develop a bit on why we'd want to avoid keeping a reference to the clipboard in the window?

I understand the code works because the `app` instance is a global reference, but the dependency of `Window` on `QApplication` remains, implicitly.

When the dependency is injected, it becomes explicit and all the interaction of the window with the clipboard happens through local variables. I thought would be more idiomatic, and I find local-only code easier to reason about YMMV).

This is what I'm taking the "more idiomatic" from:

> Explicit is better than implicit. -- The Zen of Python (`import this`)

Because I find global variables more difficult to track than local ones (leading to "complicated code" where definition are further apart form use), I wouldn't avoid dependency injection based on the simple vs complex guideline, YMMV?

> Simple is better than complex.
> Complex is better than complicated. -- The Zen of Python (`import this`)

**Comment by gonzalo-bulnes** [comment]
(rebase with no changes)

**Comment by creviera** [comment]
> @creviera Can you develop a bit on why we'd want to avoid keeping a reference to the clipboard in the window?
> 
> I understand the code works because the `app` instance is a global reference, but the dependency of `Window` on `QApplication` remains, implicitly.
> 
> When the dependency is injected, it becomes explicit and all the interaction of the window with the clipboard happens through local variables. I thought would be more idiomatic, and I find local-only code easier to reason about YMMV).
> 
> This is what I'm taking the "more idiomatic" from:
> 
> > Explicit is better than implicit. -- The Zen of Python (`import this`)
> 
> Because I find global variables more difficult to track than local ones (leading to "complicated code" where definition are further apart form use), I wouldn't avoid dependency injection based on the simple vs complex guideline, YMMV?
> 
> > Simple is better than complex.
> > Complex is better than complicated. -- The Zen of Python (`import this`)
Yes, I can expand. Qt is designed so that the `QApplication` object is always available in the application, which means the clipboard is always available. The dependency of a `QMainWindow` on a `QClipboard` is unnecessary, so there is no need for dependency injection. The `clear_clipboard` method could actually just be a helper function in the `Window` class, or the `Controller` class (probably a better place). So my reasoning for my suggestion is that it's simpler, it works with how Qt is designed, and it's not creating unnecessary dependencies or references.

**Comment by gonzalo-bulnes** [comment]
(rebase with no changes now that `main` build is green)

**Comment by gonzalo-bulnes** [comment]
I believe that Qt resorts to using a global variable for `app = QApplication([])` in order to ensure that the reference is not lost during the runtime of the application, and prevent unwanted garbage collection of underlying C++ objects.

That doesn't mean we have to rely on the fact that the variable is global, and I don't personally find that the advantages of relying on global variables are worth the drawbacks. That being said, the goal of this PR is to improve testing, and that doesn't require dependency injection, so I removed the changes related to that. :slightly_smiling_face:

**Review by creviera — APPROVED** [review]
lgtm

**Pull request merged**